web/src/components/vis/BoxPlot.vue

```diff
@@ -1,198 +1,258 @@
+import useAxisPlot from './use/useAxisPlot';
 
-export default {
-  mixins: [
-    axisPlot,
-  ],
+interface Row {
+  name: string;
+  values?: number[];
+  groups?: {
+    name: string;
+    color: string;
+    values: number[];
+  }[];
+}
+
+interface Group {
+  name: string;
+  values: number[];
+}
+
+// hoisted constants
+const margin = {
+  top: 40,
+  right: 20,
+  bottom: 50,
+  left: 150,
+};
+const xlabel = measurementColumnName;
+const ylabel = measurementValueName;
+const boxHeight = 20;
+const toFixed3 = (d: number) => d.toFixed(3);
+const count = (values: number[], min: number, max: number) => (
+  values.reduce((acc, v) => acc + (v >= min && v < max ? 1 : 0), 0));
+
+export default defineComponent({
   props: {
-    rows: { // {name: string, values?: number[],
-      //  groups?: {name: string, color: string, values: number[][]}
-      // }[]
-      type: Array,
+    rows: {
+      type: Array as PropType<Row[]>,
       default: () => [],
     },
     groups: {
-      type: Array,
-      required: false,
-      default: null,
+      type: Array as PropType<Group[]>,
+      default: () => [],
     },
   },
-  data() {
-    return {
-      margin: {
-        top: 20,
-        right: 20,
-        bottom: 50,
-        left: 120,
-      },
-      duration: 200,
-      ylabel: measurementColumnName,
-      xlabel: measurementValueName,
-    };
-  },
-  computed: {
-    scaleY() {
-      const { rows, dheight } = this;
-      return scaleBand()
-        .domain(rows.map((d) => d.name))
-        .range([0, dheight], 0.1);
-    },
-    scaleGroup() {
-      return scaleBand()
-        .domain(this.groups || [])
-        .range([0, this.scaleY.bandwidth()], 0.1);
-    },
-    xrange() {
-      let min = Number.POSITIVE_INFINITY;
-      let max = Number.NEGATIVE_INFINITY;
 
-      const pushValue = (v) => {
-        if (v < min) {
-          min = v;
-        }
-        if (v > max) {
-          max = v;
-        }
-      };
-      this.rows.forEach((row) => {
-        if (row.values) {
-          row.values.forEach(pushValue);
-        }
-        if (row.groups) {
-          row.groups.forEach((group) => group.values.forEach(pushValue));
+  components: { BoxplotBox },
+
+  setup(props) {
+    const svgRef = ref(document.createElement('svg'));
+    const mainRef = ref(document.createElement('div'));
+    const groupCount = computed(() => (props.groups || []).length || 1);
+    const width = ref(800);
+    const height = computed(() => (
+      (props.rows.length * boxHeight * groupCount.value)
+      + (margin.top + margin.bottom)));
+
+    const {
+      dwidth,
+      dheight,
+      axisPlot,
+    } = useAxisPlot(margin, width, height);
+
+    const xrange = computed(() => {
+      if (props.groups.length) {
+        return extent(flatMapDeep(props.rows, (r) => flatMapDeep(r.groups, (g) => g.values)));
+      }
+      return extent(flatMapDeep(props.rows, (r) => r.values || []));
+    }) as Ref<[number, number]>;
```

Comment:
@jeffbaumes this is the best I could come up with as a means of flattening in order to use `extent`.

Reply:
Yeah, I had realized the grouping would cause an unavoidable conditional for the added level of flattening. I'm good with this.